Add orphan check for server, volume and fip

bzhaoopenstack opened the pull request:
Add a new plugin into labchecker for orphan check.
This plugin's ptype is also 'nodepool', as the source data comes from nodepool

bzhaoopenstack committed:
Add orphan check for server, volume and fip

Add a new plugin into labchecker for orphan check.

wangxiyuan commented on openlabcmd/openlabcmd/plugins/nodepool/orphanresource.py:
```diff
@@ -0,0 +1,120 @@
+    def check(self):
+        self.failed = False
+        if self.reasons:
+            self.reasons.pop()
```
bug for `self.reason`

wangxiyuan commented on openlabcmd/openlabcmd/plugins/nodepool/orphanresource.py:
```diff
@@ -0,0 +1,120 @@
+                          'compute_id': elements[2],
+                          'floatingip': elements[3]}],
+                    'compute_ids': [elements[2]],
+                    'floatingips': [elements[3]]}
```
seems only `compute_id` is used below, what are other ids used for?

bzhaoopenstack replied:
Yeah, only compute_ids and floatingips are used. I will remove nodepool_compute_mapping

wangxiyuan replied:
where is `floatingips` used.

ZhengZhenyu commented on openlabcmd/openlabcmd/plugins/nodepool/orphanresource.py:
```diff
@@ -0,0 +1,120 @@
+            if elements[1] not in mapping:
+                mapping[elements[1]] = {
+                    'nodepool_compute_mapping':
+                        [{'nodepool_id': elements[0],
```
This seems never used? and the name is somehow missleading, could we just don't set this?

bzhaoopenstack replied:
correct, I will remove this.

liusheng commented on openlabcmd/openlabcmd/plugins/nodepool/orphanresource.py:
```diff
@@ -0,0 +1,120 @@
+        # neutron port.
+        get_fip_cmd = ("openstack --os-cloud %s floating ip list "
+                       "-c 'ID' -c 'Floating IP Address' -c 'Port' | "
+                       "egrep '[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-"
```
maybe such as commands should be more simple ?
`openstack floating ip list -c 'Floating IP Address' -c 'Fixed IP Address' -f value |grep None| awk '{ print $1 }'`

bzhaoopenstack replied:
Yeah, it works. Thanks.

liusheng commented on openlabcmd/openlabcmd/plugins/nodepool/orphanresource.py:
```diff
@@ -0,0 +1,120 @@
+        # know whether a in-use volume is used for any purpose.
+        get_volume_cmd = (
+            "openstack --os-cloud %s volume list -c 'ID' -c 'Status' | "
+            "egrep '[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-"
```
`openstack volume list --status available -f value -c ID`

liusheng replied:
and I want to remember, we may shouldn't treat all the volumes in `available` status as orphan volume. for some of the `cloud-provider-openstack` jobs, it will try to create volumes in cloud as PVs, when running jobs, the jobs may create volumes, and before attaching, the volume maybe in `available` status

bzhaoopenstack replied:
So, if fip is in the same case, the fip not attach a real port during test. How about add another filter? Check the created_at.

liusheng replied:
yes, so maybe what we can do is only to audit the orphan resource,  and then let operators to decide to delete or not..

liusheng commented on openlabcmd/openlabcmd/plugins/nodepool/orphanresource.py:
```diff
@@ -0,0 +1,120 @@
+        # online openlab env servers.
+        get_server_cmd = (
+            "openstack --os-cloud %s server list -c 'ID' -c 'Name' | "
+            "egrep '[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-"
```
similar to above

bzhaoopenstack replied:
Nice.

bzhaoopenstack committed:
Support created_at comparation

1. Add new rest API.
2. According timeout to verify the resource looks like a orphan
resource.
3. Fix info amendation issue

Yikun commented on openlabcmd/openlabcmd/plugins/base.py:
```diff
@@ -45,16 +37,19 @@ class Plugin(object):
 
-    def check_begin(self):
-        pass
+    @classmethod
```
-1, As I mentioned before, it's better to use ``__init__`` to create new plugin instance.

Yikun commented on openlabcmd/openlabcmd/cli.py:
```diff
@@ -315,7 +315,7 @@ def check(self):
             self._header_print(header)
             for plugin in plugins:
```
-1, I prefer below style to run plugin:
```python
for p in plugins:
    # p is the plugin class, plugin is the plugin instance
    plugin = p() 
    plugin.check_begin()
    plugin.check()
    plugin.check_end()
```
And ``check_begin`` is designed for some reserve operation for pre-check(such as, import env, create some fake data, or something), so we should use the original ``__init__`` to initialize the class.

wangxiyuan replied:
+1

Yikun commented on openlabcmd/openlabcmd/plugins/base.py:
```diff
@@ -26,16 +26,8 @@ def __init__(cls, name, bases, attrs):
-        # apply plugin logic - in this case connect the plugin to blinker signals
-        # this must be defined in the derived class
-        instance.register_signals()
+        cls.plugins.append(plugin)
```
nit, So, now the plugin is the plugin class, I guess we could change ``plugins`` to ``Plugins`` to avoid some confusion? and we can also add some note on it.

Yikun commented on openlabcmd/openlabcmd/plugins/base.py:
```diff
@@ -45,16 +37,19 @@ class Plugin(object):
     ptype = 'Base'
```
nit: Looks like all the member could be moved into ``__init__``

Yikun commented on openlabcmd/openlabcmd/plugins/nodepool/orphanresource.py:
```diff
@@ -0,0 +1,253 @@
+    ptype = 'nodepool'
+    name = 'OrphanResource'
+
+    def _get_endpoint_and_token(self):
```
question: Not sure why we could not use the cli to get the token and endpoint?

wangxiyuan replied:
use openstacksdk directly is easier and simpler.

Yikun commented:
FYI, the recent results [1] show that many manually created vm is found by this plugin, so it would be good add some white list mechanism into plugin.

[1] https://github.com/theopenlab/openlab/issues/314

wangxiyuan commented:
see: https://github.com/theopenlab/labkeeper/pull/171
